ggml/src/ggml-cuda/ssm_conv.cu

```diff
@@ -0,0 +1,100 @@
+#include "ssm_conv.cuh"
+
+template <int block_size>
+static __global__ void ssm_conv_f32(
+    const float * src0, const float * src1,
+    const int src0_nb0, const int src0_nb1, const int src0_nb2,
+    const int src1_nb1,
+    float * dst,
+    const int dst_nb0, const int dst_nb1, const int dst_nb2,
+    const int nc, const int ncs, const int nr) {
+
+//    const int row = blockIdx.x*blockDim.y + threadIdx.y;
+    const int tid = threadIdx.x;
+    const int i2 = blockIdx.x;
+    const int i3 = threadIdx.y;
+
+    const int ith = tid;
+    const int nth = WARP_SIZE;
+
+    // rows per thread
+    const int dr = (nr + nth - 1)/nth;
+
+    // row range for this thread
+    const int ir0 = dr * ith;
+    const int ir1 = min(ir0 + dr, nr);
+    const int ir  = ir1 - ir0;
+    
+    // {d_conv - 1 + n_t, d_inner, n_seqs}
+    // sliding window
+    const float * s = (const float *) ((const char *) src0 + ir0*src0_nb1 + i2*src0_nb0 + i3*src0_nb2); // {d_conv, d_inner, n_s}
+    const float * c = (const float *) ((const char *) src1 + ir0*src1_nb1); // {d_conv, d_inner}
+    float * x = (float *) ((char *) dst + ir0*dst_nb0 + i2*dst_nb1 + i3*dst_nb2); // {d_inner, n_t, n_s}
+    // TODO: transpose the output for smaller strides for big batches?
+    // d_inner
+    #pragma unroll
+    for (int i1 = 0; i1 < ir; ++i1) {
+        // rowwise dot product
+        // NOTE: not using ggml_vec_dot_f32, because its sum is in double precision
+        float sumf = 0.0f;
+        #pragma unroll
+        for (int i0 = 0; i0 < nc; ++i0) {
+            sumf += s[i0 + i1*ncs] * c[i0 + i1*nc];
```

Comment:
Ideally, this could use the fact that this is operating on a self-overlapping view, so advancing with `i2` shifts the view by one column.

In practice, for Mamba (and Mamba-2) models, `nc` is always 4, which might help with unrolling.

To coalesce memory accesses (at least for large prompts), I guess each warp could operate on `WARP_SIZE/nc` steps at a time over `i2`, assuming the `WARP_SIZE` is a multiple of 4 (is that always the case?), but this might need special handling of cases where `i2` is not evenly divided by that.

I don't have much experience with CUDA (yet), so this might be misleading, but hopefully this still helps.

Reply:
> Unless I'm missing something the memory access pattern here is bad with each thread accessing completely different data. You will achieve orders of magnitude higher memory bandwidth by accessing the data in a coalesced manner.

"The current memory access pattern is more suitable for CPUs. I'm thinking about ways to address this issue."

Reply:
> Ideally, this could use the fact that this is operating on a self-overlapping view, so advancing with `i2` shifts the view by one column.
> 
> In practice, for Mamba (and Mamba-2) models, `nc` is always 4, which might help with unrolling.
> 
> To coalesce memory accesses (at least for large prompts), I guess each warp could operate on `WARP_SIZE/nc` steps at a time over `i2`, assuming the `WARP_SIZE` is a multiple of 4 (is that always the case?), but this might need special handling of cases where `i2` is not evenly divided by that.
> 
> I don't have much experience with CUDA (yet), so this might be misleading, but hopefully still helps.

I’ve found a simple implementation for ssm_conv that can coalesce memory accesses, and I’ve already submitted the PR!For the ssm_scan, I'm feeling at a loss for optimization ideas."、

Reply:
>I'm feeling at a loss for optimization ideas.

Use one warp per iteration of `nc` with each thread calculating a partial sum, then combine the partial sums via `warp_reduce_sum` and have the first thread in the warp write back the result.

Reply:
Ha ha, I once also thought about using the wrap level api to calculate the sum. However, after taking a closer look, I realized that these additions are for the sum of single thread registers. Therefore, wrap_reduce_sum might not be applicable here.Thx u review.